### ISSUE TITLE
[RNN] LSTM with TimeDistributed layer converts successfully but fails when invoking

### ISSUE THREAD

**Issue by WindQAQ** [tensorflow/tensorflow#46809]
### 1. System information

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Colab
- TensorFlow installation (pip package or built from source): pip
- TensorFlow library (version, if pip package or github SHA, if built from source): tf-nightly

### 2. Code

https://colab.research.google.com/drive/1zgiZN6K1YsT70w-uWshBQA2ESuHSASn3?usp=sharing

### 3. Failure after conversion

- Model fails when invoking interpreter.
- When set `batch_size` to 1 during building model, everything works. See `with_converted_lstm.tflite` for example.

**Comment by abattery** [comment]
@renjie-liu could you take a look at this?

**Comment by abattery** [comment]
@haozha111 could you take a look at this? The problematic concatenation operator is being inserted by LowerStaticTensorList pass.

If it is an invalid transformation, it would be better to use flex fallback instead.

**Comment by haozha111** [comment]
I run the colab and check the converted tflite model, it contains the unidirectional lstm kernel, which suggests that the lower static tensorlist isn't the issue here. The converter will always try to convert to fuse kernels, and only convert to unfused if it fails to do so. Assigning the issue to @renjie-liu to take a further look.

**Comment by WindQAQ** [comment]
This should be related to

https://github.com/tensorflow/tensorflow/blob/b757b8a8db085408829fb991042fa22bc89d7a8f/tensorflow/compiler/mlir/lite/transforms/lower_static_tensor_list.cc#L402-L419

IR:
```
%0 = "tf.Const"() {value = dense<[-1, 3]> : tensor<2xi32>} : () -> tensor<2xi32>
%1 = "tf.Const"() {value = dense<[-1, 4]> : tensor<2xi32>} : () -> tensor<2xi32>
%2 = "tf.Const"() {value = dense<1.000000e+00> : tensor<f32>} : () -> tensor<f32>
%3 = "tf.Const"() {value = dense<-1> : tensor<1xi32>} : () -> tensor<1xi32>
%4 = "tf.Const"() {value = dense<0> : tensor<1xi32>} : () -> tensor<1xi32>
%5 = "tf.Const"() {value = dense<1> : tensor<1xi32>} : () -> tensor<1xi32>
%6 = "tf.Const"() {value = dense<[1, 0, 2]> : tensor<3xi32>} : () -> tensor<3xi32>
%7 = "tf.Const"() {value = dense<0> : tensor<i32>} : () -> tensor<i32>
%8 = "tf.Const"() {value = dense<-1> : tensor<i32>} : () -> tensor<i32>
%9 = "tf.Identity"(%2) {device = ""} : (tensor<f32>) -> tensor<f32>
%10 = "tf.Transpose"(%arg0, %6) {device = ""} : (tensor<?x2x3xf32>, tensor<3xi32>) -> tensor<2x?x3xf32>
%11 = "tf.Shape"(%10) {device = ""} : (tensor<2x?x3xf32>) -> tensor<3xi32>
%12 = "tf.StridedSlice"(%11, %4, %5, %5) {begin_mask = 0 : i64, device = "", ellipsis_mask = 0 : i64, end_mask = 0 : i64, new_axis_mask = 0 : i64, shrink_axis_mask = 1 : i64} : (tensor<3xi32>, tensor<1xi32>, tensor<1xi32>, tensor<1xi32>) -> tensor<i32>
%13 = "tf.TensorListReserve"(%1, %12) {device = ""} : (tensor<2xi32>, tensor<i32>) -> tensor<!tf.variant<tensor<?x4xf32>>>
```

`element_shape` of `TensorListReserve` contains -1 and is rewritten to 1, but that -1 should be `batch_size * lstm_units` instead.

**Comment by haozha111** [comment]
https://github.com/tensorflow/tensorflow/blob/b757b8a8db085408829fb991042fa22bc89d7a8f/tensorflow/compiler/mlir/lite/transforms/lower_static_tensor_list.cc#L425

From the code here, we only rewrite the first dimension in `element_shape` from -1 to 1, so it won't affect other dimensions in the `element_shape`.

**Comment by WindQAQ** [comment]
I think that -1 is exactly the first dimension?

**Comment by haozha111** [comment]
> I think that -1 is exactly the first dimension?

the element_shape is [-1, 4], and this piece of code changes it to [1, 4], which assumes the batch dim is 1.

In your model code:
inputs = tf.keras.Input(shape=(10, 2, 3))
lstm = tf.keras.layers.LSTM(4)
outputs = tf.keras.layers.TimeDistributed(lstm)(inputs)
model = tf.keras.Model(inputs, outputs)

I see a few potential issues:
1) `tf.keras.layers.LSTM` expects input shape of [batch, timesteps, features], then you applied `TimeDistributed` layer on top of this LSTM layer. According to https://www.tensorflow.org/api_docs/python/tf/keras/layers/TimeDistributed, this doesn't look quite right to me. Since `tf.keras.layers.LSTM` already process for each time step, and then you apply TimeDistributed layer again.
2) Could you explain what's the dimension in your input shape (10, 2, 3)? IIUC, 10 is the time steps, but what are 2, 3 for? Isn't LSTM usually only expects only one feature dimension?

**Comment by WindQAQ** [comment]
Sorry that it is not a good example. The usage is extracted from a private repo, and is simplied to common used API. Unfortunately, it does not match the example I post here, and I'm afraid that I cannot give more details. However, it basically contains a nested `tf.map_fn`  and some reshape ops are applied resulting in the fact that first dimension is not batch size.

**Comment by haozha111** [comment]
Thanks. If you could provide a very simple repro of the `tf.map_fn` issue, that will be great. Otherwise you could refer to this post to see if it can resolve your issue:
https://github.com/tensorflow/tensorflow/issues/40221#issuecomment-656289196

(Temporarily closed because we can't exactly reproduce the issue you have, please feel free to reopen if you find a good repro case, thanks).
